lib/puppet/http/resolver/serverlist.rb

```diff
@@ -0,0 +1,37 @@
+class Puppet::HTTP::Resolver::ServerList < Puppet::HTTP::Resolver
+  def initialize(client, server_list: Puppet[:server_list], default_port: Puppet[:masterport])
+    @client = client
+    @server_list = server_list
+    @default_port = default_port
+  end
+
+  def resolve(session, name, ssl_context: nil, &block)
+    success = false
+    @server_list.each do |server|
+      host = server[0]
+      port = server[1] || @default_port
+      uri = URI("https://#{host}:#{port}/status/v1/simple/master")
+      service = Puppet::HTTP::Service.create_service(@client, name, host, port)
+      if check_connection?(uri, session, service, ssl_context: ssl_context)
+        success = true
+        return service
+      end
+    end
+
+    raise Puppet::Error, _("Could not select a functional puppet master from server_list: '%{server_list}'") % { server_list: Puppet.settings.value(:server_list, Puppet[:environment].to_sym, true) } unless success
+  end
+
+  def check_connection?(uri, session, service, ssl_context: nil)
+    response = @client.get(uri, ssl_context: ssl_context)
+    return true if response.success?
+
+    Puppet.debug(_("Puppet server %{host}:%{port} is unavailable: %{code} %{reason}") %
+                 { host: host, port: port, code: response.code, reason: response.message })
+    return false
+  rescue => detail
+    session.add_exception(detail)
+    #TRANSLATORS 'server_list' is the name of a setting and should not be translated
+    Puppet.debug _("Unable to connect to server from server_list setting: %{detail}") % {detail: detail}
+    return false
+  end
```

Comment:
To avoid confusing this method with the superclass' version, how about renaming the method and dropping `service` from the parameter list (as we only need the `uri` and `ssl_context`).